networking.md

```diff
@@ -0,0 +1,88 @@
+
+In order to achieve that every node should aim for "future usefulness of peers" as outline below. However, as a general rule, if a node acts in violation of a `must` rule, any peer aware of that is totally in their right to drop any connection to the node without further warning. It may also keep a record of these offenses ban the peer from connecting again, if it finds them to continue offending. Such ban may be imposed upon - in increasing order - the peerId, on the IP+Port or finally the IP itself. One must keep in mind that IP addresses still rotate and thus every implementation must add a maximum timeout should they impose a ban on them, while PeerIDs can also be banned indefinitely.
+
+### Optimise for usefulness of peers
+
+A more sophisticated system to manage a healthy peer set is through optimising for usefulness of that peer to the node. in this system a node tracks all incoming messages and their costs in relation to the value it provided to them and records them as value any particular node provides to it - called "reputation". These calculations may also take into account time that passed or whether the other node would have to know about the uselessness of a particular message. Sending a useless message, in this system, is not understood as a hard violation, but as impolite behaviour and can be recorded as such. Thus creating a ranking by usefulness among the peer set.
+
+An example would be that, continuously receive a pubsub message from a node after we've forwarded it to them prior. This isn't a hard failure, as though they aren't supposed to do that, we can't know if that is because of a faulty implementation or because of network delays queuing their message. However, we'd still note this as impolite and for every time this happens deduct from their reputation as it isn't useful to us.
+
+A node attempts to always hold a certain amount of connections ("slots") to the network -- we recommend 25-50 on an on-the-shelf system. Whenever a new node connects, it can check their previously stored reputation or assign a default value and if that reputation is higher than the lowest currently connected nodes, may replace that connection (and thus drop the lowest quality connection) or otherwise refuse to take that connection.
+
+In this system a node might also only record a strong decrease in reputation but not drop a connection to a peer directly even upon a strong violation, because the node may still be more useful than others. It is better to stay connected to a crappy network than to no network. However, this doesn't not free the node from still adhering to the spec itself - it should not forward said violation or its connection might still be rightously dropped. 
+
+We also recommend to regularly check the ranking and drop and clear up the lowest 10% of slots, leaving 5% open for incoming connections and fill up the other 5% by connecting to other nodes it can find through peer discovery.
```

Comment:
I would pull this notion into an above section: it's a good heuristic even without the notion of "usefulness" and is worth duplicating ie, "We recommend nodes regularly diversify their peerSet by dropping a section of their peers (e.g. 10%) and making new connections."

Reply:
I was thinking that, too, but I am missing the background to consider it a "good heuristic". I only know this idea of dropping these connection from papers that have some form of ranking or determination of what the "lowest x%" entails. I know of "randomly drop x% of connections" only in the context of brahms, but that is a much lower number and has a much more complex structure to ensure this doesn't degrade the network over time (like certain minimal numbers of connections to always hold and a local network view to draw from).

I am not sure that generally dropping 10% of connections and _attempting_ to fill these slots with new connections (as you aren;'t guaranteed they will succeed) is not - over time - leading to you having just less and less connections until you dropped the last _good_ connection and then ... are on your own ...